Add support for specific JOSE header types

lmm-git opened the pull request:
Check https://www.rfc-editor.org/rfc/rfc7519#section-5.1 and https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.9 for more information on when the typ header should be set and to which value.

This commit allows to skip the check altogether as warranted by RFC 7515 the use of the header is optional but if set, according to RFC 7519, it should be set to "JWT". For other token types this is different (see RFC 9068).

Resolves #160

ramosbugs commented on src/verification/mod.rs:
```diff
@@ -118,6 +118,7 @@ pub(crate) struct JwtClaimsVerifier<'a, K>
 where
     K: JsonWebKey,
 {
+    allowed_jose_types: Option<Vec<JsonWebTokenType>>,
```
based on my comment in #160, let's just change this to `is_typ_check_enabled: bool`

ramosbugs replied:
since these should be unique, let's use a `HashSet` (following the example of `allowed_algs`)

ramosbugs reviewed (commented):
thanks for the PR, and for adding tests!

ramosbugs commented:
Superceded by #175

lmm-git commented:
Hm, this is a bit disappointing as I waited for your review for three weeks :(

Also the other MR does not cover the needs for verifying other types correctly.

ramosbugs commented:
Hey @lmm-git, I saw my unresolved comment about `is_typ_check_enabled` above but missed that I had agreed to merge option 3 in #160. Apologies for the miscommunication and delay (have a newborn at home as of ~3 weeks ago). I'll reopen this PR and the associated issue.

lmm-git commented:
Thanks for your response!

How do we want to proceed now? My proposal would be to revert the commit from #175 on this branch and apply the already existing commit. Or do you like it more if I rebase on top of the current main?

ramosbugs commented on src/verification/mod.rs:
```diff
@@ -181,6 +187,27 @@ where
         self
     }
 
+    /// Allows setting specific jose types, the verifier will check against during verification.
```
nit: JOSE

ramosbugs commented on src/verification/mod.rs:
```diff
@@ -203,14 +231,33 @@ where
-                )));
+            if let Some(allowed_jose_types) = &self.allowed_jose_types {
+                // Check according to https://tools.ietf.org/html/rfc7515#section-4.1.9
+                // See https://tools.ietf.org/html/rfc2045#section-5.1 for a the full Content-Type Header Field spec.
```
nit: `for a the` -> `for the`

ramosbugs commented on src/verification/mod.rs:
```diff
@@ -569,6 +616,18 @@ where
     }
 
+    /// Specifies which JSON Web Token Header types are allowed
+    pub fn set_allowed_jose_types(mut self, types: Vec<JsonWebTokenType>) -> Self {
```
let's use `IntoIterator` as with `set_allowed_algs`

ramosbugs reviewed (commented):
I've gone ahead and reverted the other PR, so this should apply cleanly against the latest commit now. just some minor comments and then I'd be happy to merge.

will118 commented:
Apologies, I misunderstood the status of this PR. I thought it had stalled on the review comment. 

Looking forward to this being merged.

lmm-git commented:
I addresses your comments and fixed some more typos in ddf65a7f2eeabd658786e50f0e634e25d03952d1

With 03f021eda6c8c23a6d7e817073a5bfa0963b0b9e I introduced a new type of a normalized json web token type. I am really not sure whether it makes the code better, but some of my reasons are:

* De-couple the parsing of the type field from the verification
* It is clear whether some function handles a normalized type value or not without any runtime penalties
* Some can use the parsing not only during verification

However, I see some reasons why not to include this commit:

* It creates a new type, which type is uncommon for this crate
* It makes understanding the code a bit more complicated
* Users might create normalized objects containing non normalized values (I kept this possibility to allow be able to use the `new_type` macro and to allow creating these values from constants)

So what do you think? I am completely fine with just leaving out the second commit. It was just an experiment of me whether it makes the code better (and I am really not sure whether it does or not)

ramosbugs commented on src/jwt/mod.rs:
```diff
@@ -20,11 +20,98 @@ new_type![
 ];
 
+#[derive(Error, Debug)]
+pub enum JsonWebTokenTypeNormalizeError {
```
let's make this `#[non_exhaustive]` to minimize future breaking changes

ramosbugs commented on src/jwt/mod.rs:
```diff
@@ -20,11 +20,98 @@ new_type![
+    fn try_from(t: &JsonWebTokenType) -> Result<NormalizedJsonWebTokenType, Self::Error> {
+        let lowercase_jwt_type = t.0.to_lowercase();
+        if lowercase_jwt_type.contains('/') {
+            if lowercase_jwt_type.starts_with("application/") {
```
where does the spec require MIME types to be `application/*`? I see where that's implied when omitted, but I don't see why some custom app couldn't have its own complex MIME type with a different top-level type

lmm-git replied:
good catch!

I changed the logic to match the standard now (hopefully ^^).

Also added some tests.

ramosbugs commented on src/jwt/mod.rs:
```diff
@@ -20,11 +20,98 @@ new_type![
+new_type![
+    /// JSON web token type field (typ)
+    ///
+    /// This types stores the raw (deserialized) value.
```
this is the normalized value, not the raw one, right?

ramosbugs commented on src/jwt/mod.rs:
```diff
@@ -20,11 +20,98 @@ new_type![
+];
+
+new_type![
+    /// JSON web token type field (typ)
```
`web token` -> `Web Token`

ramosbugs commented on src/jwt/mod.rs:
```diff
@@ -20,11 +20,98 @@ new_type![
+    /// This types stores the raw (deserialized) value.
+    ///
+    /// To compare two different JSON Web Token types, please use the normalized version via [`JsonWebTokenType::normalize`]
     #[derive(Deserialize, Hash, Ord, PartialOrd, Serialize)]
```
I'm noticing it's odd that `PartialOrd` and `Ord` are implemented but `PartialEq` and `Eq` aren't. If we don't want users to compare these non-normalized types to one another, then we should probably remove `PartialOrd`, `Ord`, and `Hash` too (not a breaking change since this type wasn't public before)

lmm-git replied:
I think they are implemented through https://github.com/ramosbugs/openidconnect-rs/blob/02525323bfdcef1c0da17835fa0d9738db354765/src/macros.rs#L161

But I removed `PartialOrd` and `Ord`.

Generally I also think that comparing these values shouldn't be (easily) possible as this might lead to wrong results. The more interesting derive would therefore be `Eq`, but as far as I can see the macro does not support switching off that specifically.

ramosbugs commented on src/jwt/mod.rs:
```diff
@@ -20,11 +20,98 @@ new_type![
 new_type![
+    /// JSON web token type field (typ)
+    ///
+    /// This types stores the raw (deserialized) value.
```
`types` -> `type`

ramosbugs commented on src/jwt/mod.rs:
```diff
@@ -20,11 +20,98 @@ new_type![
+}
+
 new_type![
+    /// JSON web token type field (typ)
```
`web token` -> `Web Token`

lmm-git replied:
There is one more "web token" in this file. Should I fix this too (maybe in an extra commit)?

ramosbugs commented on src/jwt/mod.rs:
```diff
@@ -20,11 +20,98 @@ new_type![
+    ///
+    /// This types stores the raw (deserialized) value.
+    ///
+    /// To compare two different JSON Web Token types, please use the normalized version via [`JsonWebTokenType::normalize`]
```
(period at the end)

ramosbugs commented on src/jwt/mod.rs:
```diff
@@ -20,11 +20,98 @@ new_type![
+    /// # use openidconnect::{NormalizedJsonWebTokenType, JsonWebTokenType};
+    /// let token_type: NormalizedJsonWebTokenType = JsonWebTokenType::new("jwt+at".to_string())
+    ///     .normalize()
+    ///     .unwrap();
```
I'd rather not include `.unwrap()` in the documentation snippet since it encourages users not to properly handle errors. can we wrap it in a (hidden using `#`) function that returns a `Result` and then change this to `.normalize()?`?

lmm-git replied:
I am actually not sure what you mean. I removed the `unwrap` from shown documentation at least.

ramosbugs commented on src/jwt/mod.rs:
````diff
@@ -20,11 +20,98 @@ new_type![
+    /// assert_eq!(*token_type, "application/jwt+at")
+    /// ```
+    ///
+    /// When you really want to create this type directly, please check the following items:
````
this really suggests that we shouldn't allow direct construction (without `normalize` or `TryFrom`), which means not using `new_type!`. it shouldn't be too bad to implement this type manually though (without the `new` function): https://github.com/ramosbugs/openidconnect-rs/blob/fd404985ef6c8e546f951191f4e1bc791615f5ca/src/macros.rs#L122-L146

ramosbugs reviewed (commented):
Types are a great way to capture this sort of thing, so I like this approach!

ramosbugs commented on src/jwt/mod.rs:
```diff
@@ -20,11 +20,113 @@ new_type![
+/// Error type used when normalizing [`JsonWebTokenType`] objects
+#[derive(Error, Debug)]
+#[non_exhaustive]
+#[allow(clippy::enum_variant_names)] // new error types might not start with InvalidMime
```
now that this is up to 3 variants that are very similar, I think it may be preferable to have a single error type that doesn't distinguish between the different reasons for the MIME type to be invalid. e.g.,
```rust
#[derive(Error, Debug)]
#[error("Invalid JWT type: {typ}")]
pub struct InvalidJsonWebTokenTypeError {
  typ: String,
}
```
this also includes the offending type, which should help users debug the issue. I'm assuming it'll be pretty obvious to users why the printed MIME type isn't valid 🤷‍♂️

lmm-git replied:
I thought it might be nice to allow error differentiation here but I see that it is much more simple without it.

ramosbugs commented on src/verification/mod.rs:
```diff
@@ -140,6 +141,15 @@ where
+            allowed_jose_types: Some(HashSet::from([
+                JsonWebTokenType::new("application/jwt".to_string())
+                    .normalize()
+                    .unwrap(), // used by many IdP, but not standardized
```
since this is the production code path, let's use `.expect("application/jwt should be a valid JWT type")` instead of `.unwrap()` so that the panic message is at least somewhat informative if our assumptions are wrong

ramosbugs commented on src/verification/mod.rs:
```diff
@@ -140,6 +141,15 @@ where
+                    .unwrap(), // used by many IdP, but not standardized
+                JsonWebTokenType::new("application/jose".to_string())
+                    .normalize()
+                    .unwrap(), // standard as defined in https://tools.ietf.org/html/rfc7515#section-4.1.9
```
`.expect(...)`

ramosbugs commented on src/verification/tests.rs:
```diff
@@ -47,16 +100,125 @@ fn test_jose_header() {
+                .expect("failed to deserialize"),
+            )
+            .expect("JWT type should be allowed but is not");
+    }
```
let's add one more case with a parameterized MIME type (i.e., with a `;`) since those should be accepted here

ramosbugs replied:
have you had a chance to add this test case?

lmm-git replied:
Ah sorry, I did not read exactly and thought you meant another case. Now a test case should be added.

ramosbugs reviewed (commented):
thanks for addressing the comments! almost ready to merge :-)

ramosbugs commented:
Thanks!

lmm-git commented:
Thank you for your extensive reviews!
